packages/dds/tree/docs/wip/Schema.ts

```diff
@@ -0,0 +1,621 @@
+     * - Allow replacing the whole set with a structural type for terminal / non-tree data, and use this as a replacement for values on the tree nodes.
+     * - Allow expression structural constraints for child trees, for example requiring specific traits (ex: via TreeSchema), instead of by type.
+     * There are two ways this could work:
+     *      - Constrain the child nodes based on their shape: this makes schema safe editing difficult because nodes would incur extra editing constraints to prevent them from going out of schema based on their location in such a field.
+     *      - Constrain the types allowed based on which types guarantee their data will always meet the constraints. Care would need to be taken to make sure this is sound for the schema updating mechanisms.
+     */
+    readonly types?: ReadonlySet<TreeSchemaIdentifier>;
+}
+
+export interface TreeSchema {
+    /**
+     * Schema for fields with keys scoped to this TreeSchema.
+     *
+     * This refers to the FieldSchema directly (as opposed to just supporting FieldSchemaIdentifier and having a central FieldKey -> FieldSchema map).
+     * This allows os short friendly field keys which can ergonomically used as field names in code.
+     * It also interoperates well with extraLocalFields being used as a map with arbitrary data as keys.
+     */
+    readonly localFields: ReadonlyMap<LocalFieldKey, FieldSchema>;
+
+    /**
+     * Schema for fields with keys scoped to the whole document.
+     *
+     * Having a centralized map indexed by FieldSchemaIdentifier can be used for fields which have the same meaning in multiple places,
+     * and simplifies document root handling (since the root can just have a special `FieldSchemaIdentifier`).
+     */
+    readonly globalFields: ReadonlySet<GlobalFieldKey>;
+
+    /**
+     * Constraint for local fields not mentioned in `localFields`.
+     *
+     * Allowing extraFields enables using trees like a map without having to insert localFields for all the keys.
+     *
+     * This implementation currently does not allow extra fields that are globalFields:
+     * nothing prevents adding support for this,
+     * either as part of these extra fields (and requiring them to meet this FieldSchema as well as their own),
+     * or as a separate construct.
+     * Support for them is simply omitted as it is not required by the use-cases in of this example implementation.
+     *
+     * Usually `Multiplicity.Value` should NOT be used here since no nodes can ever be in schema are in schema if you use `Multiplicity.Value` here (that would require infinite children).
+     */
+    readonly extraLocalFields: FieldSchema;
+
+    /**
+     * If true, GlobalFieldKeys other than the ones listed above in globalFields may be used to store data on this tree node.
+     * Such fields must still be in schema with their global FieldSchema.
+     */
+    readonly extraGlobalFields: boolean;
```

Comment:
I am assuming that 'augmentation' is implemented via globalFieldKeys.

If so, setting this to 'false' would prevent augmentation.  Is that something we want to support?

Reply:
Yes, globalFieldKeys would be my recommended pattern for implementing augmentations.

This allows applications to opt into augmentations anywhere which I think is what we want, but we could allow them everywhere unconditionally or nowhere and do augmentations a different way. I'd updated the documentation to cover this.